### ISSUE TITLE
Remove use of `libnpymath` and `libnpyrandom`

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#17498]
We should get this done for `1.11.0`. Using static libraries from `numpy` is unhealthy. Let me summarize the main problems and link some related issues.

From https://github.com/scipy/scipy/pull/17481#issuecomment-1328263356:
_The main issue is using a static library at all. This is bad, mainly for these two reasons:_
1. _It is not guaranteed that `numpy` and `scipy` are built with the same compilers. And that leads to potential issues that are hard to predict, as we've experienced multiple times already. Especially on Windows, where we build scipy with mingw and numpy with MSVC. That was the reason why using the `vc142` instead of the `vc141` toolchain in numpy 1.21.5-1.22.1 led to complex issues and that change had to be reverted._
2. _Cross-compiling tends to break. We haven't officially supported that in the past, but even in the one case where we do it (macOS x64-64 to arm64) it goes pretty horribly wrong: we stuff the x86-64 `libnpymath.a` into the macOS scipy wheel. This silently worked, because the required symbols were already loaded by numpy itself. Now that they're being removed from `numpy`, we're all of a sudden loading them from `libnpymath.a` and they have the wrong architecture._

For more details, see https://github.com/numpy/numpy/issues/20880.

Here are some related issues where the use of `libnpmath` showed up as a problem:

- gh-13323 (contains `libnpymath.a(npy_math.o), could not parse object file`)
- Cross-compiling issue: gh-16783
- Another cross-compiling issue: https://github.com/scipy/scipy/issues/14812#issuecomment-1204871264
- Regression because scipy 1.8/1.9 wheels contain x86-64 instructions in macOS arm64 wheel: https://github.com/numpy/numpy/issues/22673

What we should do is described in https://github.com/numpy/numpy/issues/20880#issuecomment-1327171950. Basically, vendor the sources for the functions/typedefs we still need, and rebuild them as part of the SciPy build. This is probably best done by refactoring the numpy sources a bit so there's a standalone C99 compat library with only the complex typedefs, complex functions from C99 that MSVC doesn't provide, and the blocklisted functions. And then add a command-line switch to the numpy meson build to only build that library (use a meson [disabler object](https://mesonbuild.com/Disabler.html) for that). And include numpy as a [meson subproject](https://mesonbuild.com/Subprojects.html#using-a-subproject) in the scipy source tree - this is kinda like a git submodule - so we can build it as needed, and don't actually need to vendor anything.

I don't have the bandwidth to push this forward now, but should be able to do so in Q1'23.

**Comment by rgommers** [comment]
Here is a script to check what we're still using from all of numpy (not just via `libnpymath`):
```python
import sys
import subprocess
import re

pattern = 'npy_'
def find_symbols(pattern):
    result = subprocess.run(['rg', pattern], capture_output=True, text=True)
    if result.returncode != 0:
        print(result.stderrr)
        sys.exit(result.returncode)

    return set(re.findall('(' + pattern + '\w+)\s*[,|(, ]\s*', result.stdout))


symbols_lowercase = find_symbols('npy_')
symbols_uppercase = find_symbols('NPY_')

print(symbols_lowercase)
print(symbols_uppercase)
```

Running that inside `scipy/` yields:
```
$ python ../../tmp/check_npy_symbols.py 
{'npy_types', 'npy_cpack', 'npy_type', 'npy_clog', 'npy_csqrt', 'npy_double', 'npy_ulonglong', 'npy_longlong', 'npy_float', 'npy_cpow', 'npy_promote_types', 'npy_uint8', 'npy_asarray', 'npy_cabs', 'npy_int8', 'npy_clongdouble', 'npy_longdouble', 'npy_ubyte', 'npy_cdouble_from_double_complex', 'npy_ulong', 'npy_intp', 'npy_short', 'npy_int16', 'npy_int32', 'npy_csin', 'npy_hypot', 'npy_byte', 'npy_cdouble', 'npy_uint16', 'npy_int', 'npy_uint', 'npy_carg', 'npy_ccos', 'npy_complex64', 'npy_bool_wrapper', 'npy_uint32', 'npy_bool', 'npy_uint64', 'npy_int64', 'npy_long', 'npy_uintp', 'npy_cexp', 'npy_ushort', 'npy_cfloat'}
{'NPY_SHORT', 'NPY_NO_DEPRECATED_API', 'NPY_UINT16', 'NPY_ENSURECOPY', 'NPY_BITSOF_INT', 'NPY_ARRAY_WRITEBACKIFCOPY', 'NPY_OLD', 'NPY_NEIGHBORHOOD_ITER_ZERO_PADDING', 'NPY_BITSOF_SHORT', 'NPY_BEHAVED_NS', 'NPY_PREFETCH', 'NPY_ARRAY_BEHAVED_NS', 'NPY_UINT8', 'NPY_BITSOF_LONG', 'NPY_INTP_FMT', 'NPY_USHORT', 'NPY_BYTE', 'NPY_UPDATEIFCOPY', 'NPY_LIKELY', 'NPY_UNUSED', 'NPY_FLOAT', 'NPY_SIZEOF_INTP', 'NPY_LONGLONG', 'NPY_API_VERSION', 'NPY_INT32', 'NPY_ARRAY_IN_ARRAY', 'NPY_FLOAT16', 'NPY_ULONG', 'NPY_ARRAY_F_CONTIGUOUS', 'NPY_LONG', 'NPY_CDOUBLE', 'NPY_INTP', 'NPY_ARRAY_WRITEABLE', 'NPY_SIZEOF_SHORT', 'NPY_PI', 'NPY_DOUBLE', 'NPY_CLONGDOUBLE', 'NPY_TYPES', 'NPY_SIZEOF_INT', 'NPY_CFLOAT', 'NPY_MAX_INTP', 'NPY_UNLIKELY', 'NPY_ARRAY_C_CONTIGUOUS', 'NPY_BOOL', 'NPY_SIZEOF_LONG', 'NPY_C_CONTIGUOUS', 'NPY_BITSOF_LONGLONG', 'NPY_SIZEOF_LONGDOUBLE', 'NPY_ULONGLONG', 'NPY_ARRAY_UPDATEIFCOPY', 'NPY_UBYTE', 'NPY_TYPECODE_TO_SLU', 'NPY_F_CONTIGUOUS', 'NPY_LONGDOUBLE', 'NPY_VISIBILITY_HIDDEN', 'NPY_UINT', 'NPY_ARRAY_ALIGNED', 'NPY_INT'}
```

That script isn't perfect, but it's good enough. It will also show some other issues, like the use of defines that are private to numpy (e.g. `NPY_PREFETCH`) and therefore will never be defined (they could be in the future, if we do the same configure checks as numpy, but not sure we want to go that way).

**Comment by eli-schwartz** [comment]
> And include numpy as a [meson subproject](https://mesonbuild.com/Subprojects.html#using-a-subproject) in the scipy source tree - this is kinda like a git submodule - so we can build it as needed, and don't actually need to vendor anything.

Might be a bit more lightweight to have a numpymath subproject that is used by both?

**Comment by rgommers** [comment]
That does seem like an obvious thing to do, but it would be a much larger refactor I think. The problem is that the code is really mixed up rather than standalone - headers and source files are not used only to build `libnpymath.a`, but also `multiarray_umath.so` (which contains most of numpy's core functionality). `core/src/npymath` is okay, but `core/src/common` is also needed. It's worth a try perhaps, but my sense is that it'll be pretty invasive.

**Comment by tylerjereddy** [comment]
> We should get this done for 1.11.0

IIRC I merged a few PRs related to this already, if there are more things I need to review feel free to ping me directly on them.

**Comment by rgommers** [comment]
This won't get done, and it can wait till 1.12.0 - some more simplifications to make on the numpy side first.

**Comment by rgommers** [comment]
`libnpyrandom` was removed in gh-20413. We won't get to `libnpymath` for 1.14.0 anymore, so bumping the milestone.

**Comment by ev-br** [comment]
Not sure what the story is here?

There is some usage of `npy_math.h`, mainly in scipy.special, scipy.ndimage and scipy.sparse. 
`npy_cdouble` types are used internally in `cython_lapack.pxd`, not sure what is the exact include that brings them.
And there are PRs in the pipeline which will add the include to several other places of scipy.linalg.

Are we trying to avoid including `npy_math.h` now?

```
$ git grep npy_math
doc/source/release/1.10.0-notes.rst:* `#16800 <https://github.com/scipy/scipy/pull/16800>`__: MAINT: remove \`_lib/_c99compat.h\` and use C99 rather than \`npy_math.h\`...
doc/source/release/1.10.0-notes.rst:* `#17481 <https://github.com/scipy/scipy/pull/17481>`__: MAINT: special: remove more \`npy_math.h\` usage
scipy/_build_utils/src/npy_2_complexcompat.h:#include <numpy/npy_math.h>
scipy/ndimage/src/ni_fourier.c:#include <numpy/npy_math.h>
scipy/sparse/sparsetools/complex_ops.h:#include <numpy/npy_math.h>
scipy/special/_complexstuff.h:#include <numpy/npy_math.h>
scipy/special/_hyp0f1.pxd:cdef extern from "numpy/npy_math.h":
scipy/special/_wright.h:#include <numpy/npy_math.h>
scipy/special/ellint_carlson_wrap.hh:#include <numpy/npy_math.h>
scipy/special/sf_error.cc:#include <numpy/npy_math.h>
scipy/special/xsf_wrappers.h:#include <numpy/npy_math.h>
tools/check_python_h_first.py:    "numpy/npy_math.h",
```

**Comment by rgommers** [comment]
> Are we trying to avoid including `npy_math.h` now?

No, that is unnecessary - the point is to avoid linking the `libnpymath.a` static library that is shipped with `numpy`, so that we can eventually stop shipping that. Use of `npy_math.h` doesn't imply use of that static library in most cases.

**Comment by lucascolley** [comment]
ongoing discussion on NumPy mailing list but sounds like this will not be ready for 1.17.0, bumping

**Comment by lucascolley** [comment]
This issue was closed with the removal of npymath, but we still have https://github.com/scipy/scipy/tree/main/scipy/stats/libnpyrandom, right?

**Comment by rgommers** [comment]
No, this issue is complete. It was about using a static library shipped inside `numpy`, which we no longer do. As the README in that directory says:

_This is a vendored version of NumPy's sources for libnpyrandom, with only the routines necessary for use in scipy.stats included. The vendored code is pure C, it doesn't depend on either NumPy or Python._

**Comment by lucascolley** [comment]
ah okay, thanks. What is the long-term plan there, is it here to stay? I was just wanting to keep gh-21232 up-to-date.

**Comment by rgommers** [comment]
Yeah, I don't see a reason to touch it again, fine as it is now, it's a few plain C99 files.
